### PR TITLE
Removing -X flag option and autogenerating code to create Dockerversion....

### PR DESCRIPTION
**Pull request by brahmaroutu**
...go functionality

Addresses #9207

Signed-off-by: Srini Brahmaroutu srbrahma@us.ibm.com

### CODE REVIEW

**Comment by brahmaroutu** [comment]
@tianon This PR addresses the removal of -X flag with a auto generated folder as place holder for DockerVersion and possibly other code generated on the fly in the future.

File: project/make.sh [hunk 1]
```diff
@@ -99,14 +99,10 @@ if [ "$DOCKER_EXECDRIVER" = 'lxc' ]; then
...
-LDFLAGS='
-	-X '$DOCKER_PKG'/dockerversion.GITCOMMIT "'$GITCOMMIT'"
-	-X '$DOCKER_PKG'/dockerversion.VERSION "'$VERSION'"
-'
 
-if [ -z "$DEBUG" ]; then
-	LDFLAGS="-w $LDFLAGS"
-fi
+source hack/make/.dockerversion
+IAMSTATIC="true"
```
Comment (tianon) [inline project/make.sh]: Shouldn't this be set before we `source hack/make/.dockerversion`?

Could we please instead do `source "$(dirname "$BASH_SOURCE")/make/.dockerversion"` here too so it's consistent with the other usages and encodes less assumptions?

As a final thought on this line, the double quotes (`"`) here and on the next line (`LDFLAGS`) could/should just be single quotes (`'`).

File: project/make/.dockerversion [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+rm -rf autogenerated
+mkdir -p autogenerated/dockerversion
+cat > autogenerated/dockerversion/dockerversion.go<<EOF
```
Comment (tianon) [inline project/make/.dockerversion]: This ought to have a space between `.../dockerversion.go` and `<<EOF`.

File: project/make/.dockerversion [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	// AUTOGENERATED FILE; see hack/make.sh
+	package dockerversion
+	var (
+		GITCOMMIT string = "$GITCOMMIT"
+		VERSION string = "$VERSION"
+	)
```
Comment (tianon) [inline project/make/.dockerversion]: This indentation will likely compile correctly, but IMO if we're generating Go code, it should be reasonably `gofmt`ed (ie, `gofmt -s -w` on the result should be unchanged).

File: project/make/.dockerversion [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+rm -rf autogenerated
+mkdir -p autogenerated/dockerversion
+cat > autogenerated/dockerversion/dockerversion.go<<EOF
+	// AUTOGENERATED FILE; see hack/make.sh
```
Comment (tianon) [inline project/make/.dockerversion]: This file location isn't accurate -- plugging in `$BASH_SOURCE` would be more accurate here.

```bash
cat > autogenerated/dockerversion/dockerversion.go <<EOF
// AUTOGENERATED FILE; see $BASH_SOURCE
package dockerversion
...
EOF
```

File: project/make/dynbinary [hunk 5]
```diff
@@ -4,17 +4,18 @@ set -e
 DEST=$1
 
 if [ -z "$DOCKER_CLIENTONLY" ]; then
+        export IAMSTATIC="false"
```
Comment (tianon) [inline project/make/dynbinary]: This is spaces but should be a tab. :+1:

Comment (tianon) [inline project/make/dynbinary]: Also, doing the `export` here will have unintended consequences on other builds.  ie, `project/make.sh dynbinary binary` will set `IAMSTATIC=false` and then it'll stay false for building what really is a static binary.  This is why we've got a subshell below.

**Comment by brahmaroutu** [comment]
@tianon Thanks for the review. Please see that all issues are addressed now. I have tested gofmt issue manually.

**Comment by tianon** [comment]
I'm definitely still wary of the generated code, but I think this method gives us more protection against the harmful effects it had last time, so LGTM from me.

@crosbymichael what do you think (both of the concept in general and of the specific `github.com/docker/docker/autogenerated/...` namespace)?

**Comment by brahmaroutu** [comment]
@crosbymichael @unclejack  Please review this PR, This will help getting some progress to the work for #9207

**Comment by crosbymichael** [comment]
Is there any reason why you added the autogenerated pkg instead of just sticking with `dockerversion`

**Comment by brahmaroutu** [comment]
@crosbymichael 
docker version fields are set using -X flag and that does not work for gccgo. To make it a homogenous (gc and gccgo) solution the dockerversion.go file now is generated during the build.  The content of autogenerated directory will exist only at build.

**Comment by tianon** [comment]
@crosbymichael this was what we came up with both to make it obvious
everything here is  autogenerated, but especially to solve the .gitignore
issues we ran into last time any time we want to change the name or
location of the autogenerated files

**Comment by duglin** [comment]
@brahmaroutu I've heard a rumor that if you s/autogenerate/autogen/ then people may add some LGTM's  :-)

**Comment by brahmaroutu** [comment]
@duglin I have updated the code

**Comment by duglin** [comment]
@brahmaroutu excellent - thanks!

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**